patches/server/0925-Add-Offline-PDC-API.patch

```diff
@@ -0,0 +1,99 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
+Date: Sat, 9 Jul 2022 17:28:42 +0200
+Subject: [PATCH] Add Offline PDC API
+
+
+diff --git a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
+index 601f8099f74e81c17600566b3c9b7a6dd39c9bcb..bf7c29c481539d796521de725e17a58955410866 100644
+--- a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
++++ b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
+@@ -47,6 +47,22 @@ public class PlayerDataStorage {
+         }
+ 
+     }
++    // Paper start
++    public void save(java.util.UUID uuid, CompoundTag tag) {
++        if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
++        try {
++            File file = File.createTempFile(uuid + "-", ".dat", this.playerDir);
++
++            NbtIo.writeCompressed(tag, file);
++            File file1 = new File(this.playerDir, uuid + ".dat");
++            File file2 = new File(this.playerDir, uuid + ".dat_old");
++
++            Util.safeReplaceFile(file1, file, file2);
++        } catch (Exception exception) {
++            PlayerDataStorage.LOGGER.warn("Failed to save offline player data for {}", uuid, exception); // Paper
++        }
++    }
++    // Paper end
+ 
+     @Nullable
+     public CompoundTag load(Player player) {
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+index 6d2ba650f53de8a460857f1846401a20b50cc43c..527e602ed47d11442b57eefbf3566b7793aa36db 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+@@ -304,6 +304,48 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
+ 
+         return result;
+     }
++
++    @Override
++    public org.bukkit.persistence.PersistentDataContainer getPersistentDataContainer() {
++        Player player = getPlayer();
++        if(player != null) return player.getPersistentDataContainer();
++
++        CompoundTag result = getData();
++        if(result == null) return null; // never called
```

Comment:
Am I missing something ?
This if definitely is possible if a server requests an offline player e.g. based on uuid that has not joined before.

```java
final OfflinePlayer dinnerbone = Bukkit.getOfflinePlayer(UUID.fromString("61699b2e-d327-4a01-9f1e-0ea8c3f06bc6"));
final PersistentDataContainer persistentDataContainer = dinnerbone.getPersistentDataContainer();
```
would yield a null pdc instance (unless dinnerbone played on your server before) which contradicts the `@NotNull` annotation on https://jd.papermc.io/paper/1.19/org/bukkit/persistence/PersistentDataHolder.html#getPersistentDataContainer().

Reply:
Well not really i thought that the Bukkit.getOfflinePlayer call returns null for unknown player but i was wrong